### PR TITLE
fix bandit security issues detected

### PR DESCRIPTION
**Pull request by 4383**
Fix the following issues report.

Sometimes simply using the `nosec` directive to avoid false issues like subprocess and password.

The report:
```
Run started:2018-10-31 10:45:28.866822

Test results:
>> Issue: [B701:jinja2_autoescape_false] By default, jinja2 sets autoescape to False. Consider using autoescape=True or use the select_autoescape function to mitigate XSS vulnerabilities.
   Severity: High   Confidence: High
   Location: vespene/common/templates.py:18
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b701_jinja2_autoescape_false.html
17	def environment(**options):
18	    env = Environment(**options)
19	    env.globals.update({

--------------------------------------------------
>> Issue: [B701:jinja2_autoescape_false] By default, jinja2 sets autoescape to False. Consider using autoescape=True or use the select_autoescape function to mitigate XSS vulnerabilities.
   Severity: High   Confidence: High
   Location: vespene/common/templates.py:31
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b701_jinja2_autoescape_false.html
30	    if strict_undefined:
31	        env = Environment(undefined=StrictUndefined)
32	    else:

--------------------------------------------------
>> Issue: [B701:jinja2_autoescape_false] By default, jinja2 sets autoescape to False. Consider using autoescape=True or use the select_autoescape function to mitigate XSS vulnerabilities.
   Severity: High   Confidence: High
   Location: vespene/common/templates.py:33
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b701_jinja2_autoescape_false.html
32	    else:
33	        env = Environment()
34	    t = env.from_string(str)

--------------------------------------------------
>> Issue: [B101:assert_used] Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.
   Severity: Low   Confidence: High
   Location: vespene/common/variables.py:23
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html
22	            computed = plugin.compute(self.project, results)
23	            assert computed is not None
24	            results.update(computed)

--------------------------------------------------
>> Issue: [B108:hardcoded_tmp_directory] Probable insecure usage of temp file/directory.
   Severity: Medium   Confidence: Medium
   Location: vespene/config/workers.py:8
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b108_hardcoded_tmp_directory.html
7	
8	BUILD_ROOT = "/tmp/vespene/buildroot/"
9	
10	# ---------------------------------------------------------------------------
11	# all of these settings deal with serving up the buildroot.
12	
13	# to disable file serving thorugh Django you can set this to FALSE
14	
15	FILESERVING_ENABLED = True

--------------------------------------------------
>> Issue: [B106:hardcoded_password_funcarg] Possible hardcoded password: 'you should never use this for passwords! seriously, no!'
   Severity: Low   Confidence: Medium
   Location: vespene/management/commands/tutorial_setup.py:210
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b106_hardcoded_password_funcarg.html
209	            name = "tutorial-common-variables",
210	            variables = json.dumps(dict(
211	                welcome = "Welcome to the Vespene Build Tutorial",
212	                corp_database = "foo.example.com",
213	                messaging = "msg.example.com",
214	                password = "you should never use this for passwords! seriously, no!",
215	                feature_flags = dict(
216	                    auto_frobnicate = 1,
217	                    auto_defrobnicate = 0
218	                ),
219	                repo = "foo.example.com"
220	            ))

--------------------------------------------------
>> Issue: [B105:hardcoded_password_string] Possible hardcoded password: ''
   Severity: Low   Confidence: Medium
   Location: vespene/views/forms.py:299
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b105_hardcoded_password_string.html
298	        super().__init__(*args, **kwargs)
299	        self.initial['password'] = ''
300	        self.helper = FormHelper(self)

--------------------------------------------------
>> Issue: [B404:blacklist] Consider possible security implications associated with subprocess module.
   Severity: Low   Confidence: High
   Location: vespene/workers/commands.py:8
   More Info: https://bandit.readthedocs.io/en/latest/blacklists/blacklist_imports.html#b404-import-subprocess
7	import io
8	import subprocess
9	import tempfile

--------------------------------------------------
>> Issue: [B602:subprocess_popen_with_shell_equals_true] subprocess call with shell=True identified, security issue.
   Severity: High   Confidence: High
   Location: vespene/workers/commands.py:92
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b602_subprocess_popen_with_shell_equals_true.html
91	
92	    process = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=shell, env=env)
93	

--------------------------------------------------
>> Issue: [B404:blacklist] Consider possible security implications associated with subprocess module.
   Severity: Low   Confidence: High
   Location: vespene/workers/registration.py:14
   More Info: https://bandit.readthedocs.io/en/latest/blacklists/blacklist_imports.html#b404-import-subprocess
13	import fcntl
14	import subprocess
15	import os

--------------------------------------------------
>> Issue: [B607:start_process_with_partial_path] Starting a process with a partial executable path
   Severity: Low   Confidence: High
   Location: vespene/workers/registration.py:66
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b607_start_process_with_partial_path.html
65	    def guess_hostname(self):
66	        return subprocess.check_output("hostname").decode('utf-8').strip()
67	

--------------------------------------------------
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   Location: vespene/workers/registration.py:66
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b603_subprocess_without_shell_equals_true.html
65	    def guess_hostname(self):
66	        return subprocess.check_output("hostname").decode('utf-8').strip()
67	

--------------------------------------------------

Code scanned:
	Total lines of code: 5313
	Total lines skipped (#nosec): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0.0
		Low: 7.0
		Medium: 1.0
		High: 4.0
	Total issues (by confidence):
		Undefined: 0.0
		Low: 0.0
		Medium: 3.0
		High: 9.0
Files skipped (0):
```

### CODE REVIEW

File: vespene/common/templates.py [hunk 1]
```diff
@@ -28,9 +28,9 @@ def template(str, template_context, strict_undefined=False):
     Return the value of string templated through the dictionary 'template_context'
     """
     if strict_undefined:
-        env = Environment(undefined=StrictUndefined)
+        env = Environment(undefined=StrictUndefined, autoescape=True)
     else:
-        env = Environment()
+        env = Environment(autoescape=True)
```
Comment (mpdehaan) [inline vespene/common/templates.py]: This looks good, but I need to make sure the various places we use "| safe" in the templates keep working.  Will check out.

File: vespene/views/forms.py [hunk 2]
```diff
@@ -296,7 +296,7 @@ class Meta:
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.initial['password'] = ''
+        self.initial['password'] = '' # nosec
```
Comment (mpdehaan) [inline vespene/views/forms.py]: Ah, interesting.  Yeah perhaps this #nosec should be added to that previous "warn" instead.

**Comment by mpdehaan** [comment]
Hey thanks for this!

Bandit looks super interesting - I added some comments above, it looks like adding # nosec on a few will make those warnings go away, and I'll take a look at the Jinja2 escaping questions.

That's my main question about this one, because view_helpers code has some things that generate widget code, but it might be fine.

I agree /tmp/ isn't the right place, but I don't want a build root that moves on restart :) That would break fileserving and all the cleanup scripts would not know where to look for old build roots pre-restart. We do need a predictable build root, /opt/vespene_build is probably a fine default.  This would require tweaking the setup scripts too, but it wouldn't be a big deal.

**Comment by 4383** [comment]
In my other PR #52 I introduce automated bandit check inside travis, we need to discuss if we want to run bandit with or without the nosec ignore.

What's your position?

**Comment by mpdehaan** [comment]
> In my other PR #52 I introduce automated bandit check inside travis

I commented in the other one that it feels super weird to be building Vespene with another build tool, so I don't want to include a travis file.  That being said, Bandit looks interesting.

Assuming it just has a CLI, how about a makefile target to make it easy to run regularly?

Thanks very much for taking the time to fix all of those too!

File: vespene/management/commands/tutorial_setup.py [hunk 3]
```diff
@@ -205,13 +205,14 @@ def handle(self, *args, **options):
...
 
+        warn = "you should never use this for passwords! seriously, no!" # nosec
         variable_set1, _ = VariableSet.objects.get_or_create(
             name = "tutorial-common-variables",
             variables = json.dumps(dict(
                 welcome = "Welcome to the Vespene Build Tutorial",
                 corp_database = "foo.example.com",
                 messaging = "msg.example.com",
-                password = "you should never use this for passwords! seriously, no!",
+                password = warn,
```
[the remark this answers is not in the document]
Comment (4383) [reply vespene/management/commands/tutorial_setup.py]: Another approach would be ignore this file during bandit check... but personnaly I'm not fan

[the remark this answers is not in the document]
Comment (4383) [reply vespene/management/commands/tutorial_setup.py]: your comment below is already implemented in `vespene/management/commands/tutorial_setup.py`

**Comment by 4383** [comment]
@mpdehaan so do you agree the current changes (expected the time to check the view_helpers subject) ? 
My question is not to merge this PR immediately but rather than to avoid doubt.

**Comment by mpdehaan** [comment]
Yeah I need to test this first.  This will probably be early next week.

(The "warn" in the JSON does need to be removed)

**Comment by 4383** [comment]
@mpdehaan nice

**Comment by mpdehaan** [comment]
Upon review, it is clear the Jinja2 autoescaping parameters are only needed when doing web templates and this may break lots of things including commands that rely on Jinja, like plugin configurations.  I'll look about whether it is meaningful to turn it on in specific instances.

Seeing the warn change in the sample config is also untested and *WILL* break code, so I'm not going to merge this PR. That string is not a real password, but is just an instructional lesson about not putting passwords there in the real world, so it's kind of a false positive.  That seems like a "# nosec" to me as well.

A new pull request to just add the "# nosec" additions seperately would be welcome, I'll persue the autoescaping one now.

**Comment by mpdehaan** [comment]
Since this was security related I wanted to post a bit more on this ticket.

I've gone through the Jinja2 code and in quite a few places we are using helper functions to make HTML components - escaping this would break those.

What we need to do is make sure that in places where we are NOT in a form (which is the main "lists" in the UI) that any element shown in that grid is protected by html.escape() ... which I have added manually.

I don't think Jinja2 autoescape is appropriate here because we would have to add the safe filter ("| safe") to nearly all of those invocations from Jinja2, which would make the Jinja2 evaluation meaningless.

The right fix is to use html.escape() everywhere that user input could come from that is not going to be displayed in a form.

I think we're in pretty good state here right now except for *possibly* the "messages" flash area, which I'll read up on and test. I'll file a ticket for that one to make sure it's taken care of.

**Comment by 4383** [comment]
@mpdehaan ack

**Pull request closed without merging**